lib/new_relixir/poller.ex

```diff
@@ -0,0 +1,68 @@
+defmodule NewRelixir.Poller do
+  use GenServer
+  require Logger
+
+  @poll_interval Application.get_env(:new_relic, :poll_interval) || 30_000
+
+  ## API
+
+  def start_link(poll_fun, error_cb \\ &default_error_cb/2) do
+    GenServer.start_link(__MODULE__, %{poll_fun: poll_fun, error_cb: error_cb})
+  end
+
+  ## Callbacks
+
+  def init(state) do
+    timer = :erlang.send_after(@poll_interval, self(), :poll)
+    {:ok, Map.put(state, :timer, timer)}
+  end
+
+  def handle_call(_msg, _from, state) do
+    {:reply, :ok, state}
+  end
+
+  def handle_cast(_msg, state) do
+    {:noreply, state}
+  end
+
+  def handle_info(:poll, %{poll_fun: poll_fun, error_cb: error_cb, timer: old_timer}) do
+    :erlang.cancel_timer(old_timer)
+    timer = :erlang.send_after(@poll_interval, self(), :poll)
+    {:ok, hostname} = :inet.gethostname()
+    try do
+      case poll_fun.() do
+        {[], []} ->
```

Comment:
I think this rule is never getting matched. Both `webtransaction_total` and `db_total` don't return empty lists, even when there is no data. So we are making "empty" requests to New Relic. We can see it when we log the calls to `push_data`:

```
[debug] Pushing data: [595128166, 1511802221, 1511802251, [[%{name: "HttpDispatcher", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Database/all", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Errors/all", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Errors/allWeb", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Instance/Reporting", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]]]]
[debug] Pushing data: [595128166, []]
[debug] Pushing data: [595129818, 1511802251, 1511802281, [[%{name: "HttpDispatcher", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Database/all", scope: ""}, [0, 0, 0, 0, 0, 0]], [%{name: "Errors/all", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Errors/allWeb", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]], [%{name: "Instance/Reporting", scope: ""}, [0, 0.0, 0.0, 0.0, 0.0, 0.0]]]]
[debug] Pushing data: [595129818, []]
```

Reply:
One way we can fix this is adding an extra clause to `transform_aggregated_metrics`, like:

```elixir
def transform_aggregated_metrics([], [], time), do: {[], [], time}
def transform_aggregated_metrics(metrics, errors, time) do
  ms = Enum.flat_map(metrics , &transform_histogram/1)
  errs = Enum.map(errors, &transform_error_counter/1)

  {[webtransaction_total(ms), db_total(ms) | errors_total(errs) ++ ms], errs, time}
end
```

Then match it with the expected arity:

```elixir
case poll_fun.() do
  {[], [], _interval} ->
    Logger.debug "Nothing to send."
    :ok
```

Reply:
Done in 2d4623d2975046a5ff2dc01aa4e564f417e263c2